components/codec/src/bits/buffered_read.rs

```diff
@@ -0,0 +1,187 @@
+// Copyright 2023 The CeresDB Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::bits::{Bit, Error, BIT_MASKS};
+
+/// BufferedReader
+/// BufferedReader encapsulates a buffer of bytes which can be read from.
+#[derive(Debug)]
+pub struct BufferedReader<'a> {
+    /// internal buffer of bytes
+    bytes: &'a [u8],
+    /// index into bytes
+    byte_idx: usize,
+    /// position in the byte we are currently reading
+    bit_idx: u32,
+}
+
+impl<'a> BufferedReader<'a> {
+    /// new creates a new `BufferedReader` from `bytes`
+    pub fn new(bytes: &'a [u8]) -> Self {
+        BufferedReader {
+            bytes,
+            byte_idx: 0,
+            bit_idx: 0,
+        }
+    }
+
+    fn get_byte(&mut self) -> Result<u8, Error> {
```

Comment:
I guess the `self.byte_idx` is always valid, so there is no need to use `self.bytes.get` here?

And here a read-only self reference is enough here.